### PR TITLE
Add the ability to customize available models for OpenAI-compatible services

### LINKED ISSUE #11075: Implement Groq or OS models for Assistant Tab (settings.assistant)
### Check for existing issues

- [X] Completed

### Describe the feature

Currently, the Assistant Tab is limited to using the OpenAI API, which can lead to some challenges for open source projects. To provide more flexibility, I suggest adding a configurable field to change the base_url for the assistant tab. This would allow users to access free AI services like Groq, or alternatively use Ollama, LiteLLM, or other models by specifying their base_url. However, I want to note that I haven't fully vetted or tested these alternative options yet, so some additional checks may be needed before implementing this.

image

### If applicable, add mockups / screenshots to help present your vision of the feature

It would be great to have a window or dialog similar to the request panel, where users can:

1. Enter the base_url for alternative AI models/services
2. Provide names and settings for custom models 
3. Save and manage multiple configurations

The dialog could have a dropdown to select pre-defined options (OpenAI, Groq, Ollama, etc.) as well as the ability to add new custom entries with a friendly name, base_url, and required parameters.

This would allow easy switching between different AI models/providers without manually entering URLs/settings every time, providing a seamless experience for developers/users exploring various AI capabilities.
_No response_

### LINKED ISSUE #11984: [Assistant] add `custom_model` parameter
### Check for existing issues

- [X] Completed

### Describe the feature

The current assistant does not support the configuration of custom models, and there is a hope to add custom models with configurations such as:

```json
  "assistant": {
    "version": "1",
    "provider": {
      "name": "openai",
      "type": "openai",
      "default_model": "gpt-3.5-turbo",
      "custom_model": ["custom_model_1", "custom_model_2"],
      "api_url": "https://www.example.com"
    }
  }
```

### PR DESCRIPTION
**Pull request by amtoaer**
Closes #11984, closes #11075.

Release Notes:

- Added the ability to customize available models for OpenAI-compatible services ([#11984](https://github.com/zed-industries/zed/issues/11984)) ([#11075](https://github.com/zed-industries/zed/issues/11075)).

image

### CODE REVIEW

**Comment by christ-offer** [comment]
Very nice! :D 

image

File: crates/open_ai/src/open_ai.rs [hunk 1]
```diff
@@ -55,6 +59,12 @@ pub enum Model {
     #[serde(rename = "gpt-4o", alias = "gpt-4o-2024-05-13")]
     #[default]
     FourOmni,
+    #[serde(rename = "custom")]
+    Custom {
+        name: String,
+        #[serde(default = "custom_default_max_tokens")]
```
Comment (rtfeldman) [inline crates/open_ai/src/open_ai.rs:65]: I think we should make this a required field instead of giving it a default. We don't really have any basis for guessing what the token limit might be for an unknown model! 😄

File: crates/assistant/src/completion_provider.rs [hunk 2]
```diff
@@ -24,6 +24,21 @@ use settings::{Settings, SettingsStore};
...
+/// Choose which model to use for openai provider based on the available models and the default model.
+fn choose_openai_model(
+    model: &::open_ai::Model,
+    available_models: &Option<Vec<::open_ai::Model>>,
+) -> ::open_ai::Model {
+    let available_models = available_models.as_ref();
+
+    // available_models is set but default model isn't in the list, use the first available model instead
+    if available_models.is_some_and(|models| !models.is_empty() && !models.contains(model)) {
+        available_models.unwrap()[0].clone()
```
Comment (rtfeldman) [inline crates/assistant/src/completion_provider.rs:36]: Can we rule out potential panics here (from `unwrap` and `[0]`)?

For example, by doing something like this:

```rust
available_models
    .and_then(|models| models.find(model).or_else(|| models.first()))
    .unwrap_or_else(|| model)
    .clone()
```

File: crates/assistant/src/assistant_settings.rs [hunk 3]
```diff
@@ -163,6 +163,7 @@ pub enum AssistantProvider {
         model: OpenAiModel,
         api_url: String,
         low_speed_timeout_in_seconds: Option<u64>,
+        available_models: Option<Vec<OpenAiModel>>,
```
Comment (rtfeldman) [inline crates/assistant/src/assistant_settings.rs:166]: Have you thought about having this be `Vec<OpenAiModel>` instead? Is it important to be able to have the case `Some(vec![])` work, or does that not make sense?

**Review by rtfeldman — CHANGES_REQUESTED** [review]
Thanks for the contribution - this is looking good overall! 😃 

I have one requested design change, and then some questions about code structure.

**Comment by amtoaer** [comment]
Thank you. I've updated it. Please let me know if there are any other issues.

**Review by rtfeldman — APPROVED** [review]
Nice, thank you @amtoaer! 🎉

**Comment by sa1** [comment]
Some central documentation for new features would be nice.

**Comment by sahadev0079** [comment]
image
@christ-offer what's wrong here ?

**Comment by pramjan** [comment]
The GIF in this thread changes the displayed image too fast, making it hard to follow. Also, the longer image above is missing some information. Zed Docs mentions "language models" as the key, and so does gliptic in [16576]](https://github.com/zed-industries/zed/discussions/16576).

It's confusing!

Please provide a simple, correct method to set up OpenRouter models in Zed (one model example is enough, thank you!).

**Comment by notpeter** [comment]
> Please can someone provide a simple, correct method to set up OpenRouter models in Zed (one model example is enough, thank you!).

I believe they are using OpenAI API compatibility, so based on the [Asssistant Configuration: OpenAI](https://zed.dev/docs/assistant/configuration#openai) docs:

```json
  "language_models": {
    "openai": {
      "api_url": "https://api.deepseek.com/v1",
      "available_models": [
        {
          "provider": "openai",
          "name": "deepseek_chat",
          "max_tokens": 32000
        },
        {
          "provider": "openai",
          "name": "deepseek_code",
          "max_tokens": 328000
        }
      ]
    }
  }
}
```
Untested, but you will need to enter your API key under the OpenAI section (press enter to confirm) of the Assistant configuration ("assistant: show configuration" from the command palette).
